module/surface_sam_hps.c

```diff
@@ -1021,23 +1168,29 @@ static int shps_probe(struct platform_device *pdev)
 	if (!link)
 		goto err_devlink;
 
-	surface_sam_san_set_rqsg_handler(shps_dgpu_handle_rqsg, pdev);
+	if (detected_traits.notification_method == SHPS_NOTIFICATION_METHOD_SAN) {
+		surface_sam_san_set_rqsg_handler(shps_dgpu_handle_rqsg, pdev);
+	} else if (detected_traits.notification_method == SHPS_NOTIFICATION_METHOD_SGCP) {
+		shps_try_start_sgcp_notification(pdev, &drvdata->sgpc_handle);
+	}
```

Comment:
There should be error handling here. Also you can move the `surface_sam_san_consumer_register` part here, so that there's a single method specific setup section.

Reply:
I thought about moving the san consumer register part too, but it'll return `EPROBE_DEFER` if the san module hasn't loaded yet, and I didn't want the driver to do a lot of work here before it got to that point.

Reply:
Right, that makes sense.